### PR TITLE
Add date extensions and tests

### PR DESCRIPTION
**Pull request by jebediahelliott**
Issue #9

### CODE REVIEW

File: src/AndcultureCode.CSharp.Extensions/DateExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+        /// <summary>
+        /// Convenience method to calculate an age from a birthdate
+        /// </summary>
+        public static int CalculateAge(this DateTime birthdate)
+        {
+            var age = DateTime.Today.Year - birthdate.Year;
+            // Subtract a year if the birthday has not happened yet this year
+            // https://stackoverflow.com/questions/9/how-do-i-calculate-someones-age-in-c
+            if (birthdate > DateTime.Today.AddYears(-age)) age--;
```
Comment (wintondeshong) [inline src/AndcultureCode.CSharp.Extensions/DateExtensions.cs]: I realize this code was copied from the other project, but lets take a pass at cleaning it up to follow our best practices.

In this case, we do not exclude curly braces even for single statements.

```csharp
if (birthdate > DateTime.Today.AddYears(-age))
{
    age--;
}
```

File: src/AndcultureCode.CSharp.Extensions/DateExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        /// <summary>
+        /// Convenience method to subtract weekdays
+        /// </summary>
+        public static DateTime SubtractWeekdays(this DateTime originalDate, int weekdays)
+        {
+            DateTime tmpDate = originalDate;
+            while (weekdays > 0)
+            {
+                tmpDate = tmpDate.AddDays(-1);
+                if (tmpDate.DayOfWeek < DayOfWeek.Saturday &&
```
Comment (wintondeshong) [inline src/AndcultureCode.CSharp.Extensions/DateExtensions.cs]: Same as above, don't exclude the curly braces

File: src/AndcultureCode.CSharp.Extensions/DateExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            if (birthdate > DateTime.Today.AddYears(-age)) age--;
+            return age;
+        }
+
+        /// <summary>
+        /// Convenience method to subtract weekdays
+        /// </summary>
+        public static DateTime SubtractWeekdays(this DateTime originalDate, int weekdays)
+        {
+            DateTime tmpDate = originalDate;
```
Comment (wintondeshong) [inline src/AndcultureCode.CSharp.Extensions/DateExtensions.cs]: We want to use `var` when possible. Being `originalDate` is of type `DateTime`, assignment to `tmpDate` implies the type.

File: src/AndcultureCode.CSharp.Extensions/DateExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            }
+            return tmpDate;
+        }
+
+        /// <summary>
+        /// Convenience method to subtract weekdays
+        /// </summary>
+        public static DateTimeOffset SubtractWeekdays(this DateTimeOffset originalDate, int weekdays)
+        {
+            DateTimeOffset tmpDate = originalDate;
```
Comment (wintondeshong) [inline src/AndcultureCode.CSharp.Extensions/DateExtensions.cs]: As above, use `var`

File: src/AndcultureCode.CSharp.Extensions/DateExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            while (weekdays > 0)
+            {
+                tmpDate = tmpDate.AddDays(-1);
+                if (tmpDate.DayOfWeek < DayOfWeek.Saturday &&
```
Comment (wintondeshong) [inline src/AndcultureCode.CSharp.Extensions/DateExtensions.cs]: As above, add curly braces

File: src/AndcultureCode.CSharp.Extensions/DateExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        {
+            if (inclusive)
+            {
+                return date.IsBetweenDatesInclusive(minDate, maxDate);
+            }
+
+            return date.IsBetweenDatesExclusive(minDate, maxDate);
+        }
+
+        #region Private Methods
```
Comment (wintondeshong) [inline src/AndcultureCode.CSharp.Extensions/DateExtensions.cs]: Recommend wrapping public methods above in a region as well

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+            var testDate = DateTimeOffset.Now;
+
+            // Act
+            var result = testDate.AtMidnight();
+
+            // Assert
+            result.Hour.ShouldBe(0);
+            result.Minute.ShouldBe(0);
+            result.Second.ShouldBe(0);
+
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: Remove extra line

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        public void AtMidnight_Returns_Date_With_Time_Set_To_Midnight()
+        {
+            // Arrange
+            var testDate = DateTimeOffset.Now;
+
+            // Act
+            var result = testDate.AtMidnight();
+
+            // Assert
+            result.Hour.ShouldBe(0);
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: Might be a good place to use http://docs.shouldly-lib.net/docs/shouldsatisfyallconditions

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+        public void AtMidnight_Does_Not_Change_The_Date()
+        {
+            // Arrange
+            var testDate = DateTimeOffset.Now;
+
+            // Act
+            var result = testDate.AtMidnight();
+
+            // Assert
+            result.Year.ShouldBe(testDate.Year);
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: Recommend use of http://docs.shouldly-lib.net/docs/shouldsatisfyallconditions

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+            result.Month.ShouldBe(testDate.Month);
+            result.Day.ShouldBe(testDate.Day);
+        }
+
+        #endregion
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: While we do tend to have the name into `endregion` everywhere, it is especially helpful in test fixtures. 

So `#endregion AtMidnight`

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+            birthdate = birthdate.AddYears(-5);
+
+            // Act
+            var result = birthdate.CalculateAge();
+
+            // Assert
+            result.ShouldBe(5);
+        }
+
+        #endregion
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: Endregion label

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+
+        #region SubtractWeekdays -- DateTime
+
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: Remove extra newline

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+            var day = new DateTime(2019, 11, 16);
+
+            // Act
+            var result = day.SubtractWeekdays(5);
+
+            // Assert
+            result.DayOfWeek.ShouldBe(DayOfWeek.Monday);
+        }
+
+        #endregion
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: endregion label

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+
+        #region SubtractWeekdays -- DateTimeOffset
+
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: Remove extra newline

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+            // Act
+            var result = dateWithOffset.SubtractWeekdays(3);
+
+            // Assert
+            result.DayOfWeek.ShouldBe(DayOfWeek.Monday);
+        }
+
+        [Fact]
+        public void SubtractWeekdays_With_Offset_Returns_Correctly_When_Day_Is_In_A_Different_Week()
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: For this and other test methods in this fixture, what does 'Correctly' mean? Consider being a bit more explicit that "X" input expects "Y" output

Additionally, recommend putting `_Returns_XYZ` at the end of the test method name. So a cadence of 'given X', 'given Y', 'when Z', 'when W', 'returns T'. Often times test methods tend to 'group' themselves on the 'when' and 'given' clauses

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+            var dateWithOffset = new DateTimeOffset(day);
+
+            // Act
+            var result = dateWithOffset.SubtractWeekdays(5);
+
+            // Assert
+            result.DayOfWeek.ShouldBe(DayOfWeek.Monday);
+        }
+
+        #endregion
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: endregion label

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+            var maxDate = date.AddDays(3);
+
+            // Act
+            var result = date.IsBetweenDates(minDate, maxDate, false);
+
+            // Assert
+            result.ShouldBeFalse();
+        }
+
+        #endregion
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: endregion label

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+
+        #region IsBetweenDates -- Inclusive Default
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: Recommend incorporating these region names into the beginning of your actual test method names. Makes easier to follow and maintain. For instance...

`IsBetweenDates_InclusiveDefault_Overload_When_Date_Is_Between_Min_And_Max_Returns_True()`

Then when looking at a list of all test names, you can easily see the groupings

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+            // Act
+            var result = dateWithOffset.SubtractWeekdays(5);
+
+            // Assert
+            result.DayOfWeek.ShouldBe(DayOfWeek.Monday);
+        }
+
+        #endregion
+
+        #region IsBetweenDates -- No Default
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: Recommend incorporating these region names into the test names...

`IsBetweenDates_NoDefault_Overload_When_Date_Is_Between_Min_And_Max_When_Inclusive_Is_True_Returns_True()`

File: test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+            var maxDate = date.AddDays(3);
+
+            // Act
+            var result = date.IsBetweenDates(minDate, maxDate);
+
+            // Assert
+            result.ShouldBeTrue();
+        }
+
+        #endregion
```
Comment (wintondeshong) [inline test/AndcultureCode.CSharp.Extensions.Tests/DateExtensionsTests.cs]: endregion label

**Comment by jebediahelliott** [comment]
@wintondeshong I made those updates if you want to take another look

**Pull request merged**